### DISCUSSION TITLE
Rendering `.svg` files as components

### DISCUSSION THREAD

**Discussion by natemoo-re** [withastro/roadmap#667] [Proposal]
### Body

# Summary

Let's support importing and using `.svg` files as Astro components.

# Background & Motivation

Users have reported that `.svg` files are currently difficult to use—this is a problem we should provide an optimized solution for. Working with `.svg`s doesn’t need to be difficult or bloat your client-side JavaScript.

The Astro team has discussed and experimented with an "icon" primitive since before Astro v1.0 shipped. Now that we have shipped 3.0 with full `astro:assets` support, I firmly believe all of the pieces are in place to finally make this happen.

![svgs in astro?](https://github.com/withastro/roadmap/assets/7118177/7e12fa95-3b79-40b0-a68f-65a96dd23a83)

[TODO] Add some background about how `.svg` files work in other frameworks (spoiler alert: it's usually not easy and leads users down the wrong path)

**History**
- [`astro-icon`](https://github.com/natemoo-re/astro-icon) was built as an experimental way to support `.svg` files in Astro, before we had server rendering _or_ an integration API. Though @stramel and I have worked to get [`astro-icon` to a v1.0](https://github.com/natemoo-re/astro-icon/pull/111), I still haven't pulled the trigger and released it due to concerns over bloated server bundles.
- @FredKSchott made an early proposal for [`astro/icons`](https://github.com/withastro/roadmap/discussions/107) which had a significantly reduced scope and came with a pre-baked icon set. Feedback was that it was not flexible enough for most uses, so we punted on it.
- `astro:assets` ([RFC-0030](https://github.com/withastro/roadmap/blob/main/proposals/0030-core-image-story.md)) added support for loading image files with some metadata, but mostly optimizing bitmap images. Frustratingly, this didn't include `.svg` files, but a [community PR added pass-through behavior](https://github.com/withastro/astro/pull/7643) for `.svg` files. This shipped unflagged in 3.0.

Unfortunately, these explorations still haven't solved the problem, which is that users just want to be able to import and render an `.svg` file. `.svg` files are possible to support as components that can pass props to the root `svg` element.

Additionally, when exploring `astro-icon` v1, I stumbled onto a very promising optimization technique for embedding `.svg` files in an HTML document. Astro is uniquely positioned to take advantage of this spritesheet-inspired approach that sends minimal bytes over the network. The technique involves rendering each `.svg` into a `<symbol>` when first used on the page, and subsequently a `<use>` declaration. This is similar to a traditional spritesheet approach, but is automated without any drawbacks or need to use a Provider pattern.

# Goals

- Allow users to import and render local `.svg` files as if they were an Astro component
- Allow passing props onto the root `<svg>` element while overriding the existing props
- Automate best practices for to minimize performance footguns
  - Including many inlined `<svg>` on a page can hurt browser parsing performance. We can automatically reduce the number of nodes by using `<symbol>` and `<use>`.
  - Inline `<svg>` often have `xmlns` and `xmlns:xlink` and `version` attributes. These are not needed in HTML5 and we can automatically drop them to save a few bytes.
- Support this in a backwards-compatible way, meaning the current public interface of an `.svg` import remains unchanged
- Stretch: support rendering `.svg` components in frameworks without inlining them into the client side JavaScript
    - Creating framework components for `.svg` icons is an anti-pattern. It’s a terrible idea that bloats your JavaScript and creates all sorts of headaches for HMR and clientside performance in production.
    - If we provide a great `<Icon>` component for each supported framework, we’ll make it easy for users to avoid this anti-pattern. It’s so common now because there aren’t many other good solutions.

# Non-goals

I'm jumping ahead to the next stage, but I wouldn't consider support for [`@iconify/json`](https://www.npmjs.com/package/@iconify/json) sets a goal for this initially. There are too many gotchas with SSR and tree-shaking that we'd need to figure out. We can leave this to userland for now.

# Example

User-facing API

## Option A (original)

```astro
---
import Logo from '../assets/logo.svg'
---

<!-- first render: includes initial <symbol> -->
<Logo width={48} height={48} />

<!-- subsequent renders: automatically optimized with <use> -->
<Logo width={48} height={48} />
```

## Option B [(see comment)](https://github.com/withastro/roadmap/discussions/667#discussioncomment-6842106)

```astro
---
import { Svg } from 'astro:components'
import Logo from '../assets/logo.svg'
---

<!-- first render: includes initial <symbol> -->
<Svg src={Logo} width={48} height={48} />

<!-- subsequent renders: automatically optimized with <use> -->
<Svg src={Logo} width={48} height={48} />
```

## Framework API?

```ts
import { Svg } from '@astrojs/react/components'
import Logo from '../assets/logo.svg'

export function ReactComponent() {
  // This would only render `<use>` for maximum size reduction
  // We would automatically inject the `<symbol>` elsewhere
  return <Svg src={Logo} width={48} height={48} />
}
```

I have created a [spike branch](https://github.com/withastro/astro/compare/main...spike/svg) to explore this idea.

**Comment by Princesseuh** [comment]
How would this interact with `astro:assets`? Something that happens fairly often is someone having a list of images, that can be both svgs and raster formats and passing each entry of the list to `<Image />` (hence why the passthrough was added). I'll also note that if we ever drop Squoosh, our Sharp image service could handle some transformation for SVGs, which some users might've wanted.

Apart from those points, I love it and it seems wonderful.

**Reply by natemoo-re** [reply]
A great question! I was originally thinking of this as tangential to the existing `astro:assets` features, but if there's a world where `astro:assets` can also optimize SVGs, then we could certainly factor that in.

Initial plan was that the `import Component from '../file.svg'` shape would remain unchanged. Instead of a plain object, it would actually be a function that has the `ImageMetadata` assigned to it. My current spike is building the component logic at compile-time without the `astro:assets` optimization, but we could probably construct the component from the optimized result at runtime as well.

**Comment by natemoo-re** [comment]
One possible alternative API is more like the existing `<Image src={img} />` component from `astro:assets`. Instead of rendering `<img src="*.svg">`, it would inline the `.svg` directly.

```astro
---
import { Svg } from 'astro:components'
import Logo from '../assets/logo.svg'
---

<!-- first render: includes initial <symbol> -->
<Svg src={Logo} width={48} height={48} />

<!-- subsequent renders: automatically optimized with <use> -->
<Svg src={Logo} width={48} height={48} />
```

Bikeshedding possible names:
- `Svg` seems obvious. Casing is kinda weird, but `SVG` seems weirder?
- `Icon`. Makes sense but `.svg` files aren't always icons.
- `Embed`. Not the same as [`<embed>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/embed).
- `Inline`. Overloaded. Not the same a `inline` vs `block` in CSS.
- `InlineSvg`. Literal but verbose.

Pros:
- Alignment. This API closely mirrors the existing `Image` component.
- Consistency. This same API could be used in framework integrations (React, etc) as well.
- Compatability. This API could potentially consume third-party (iconify) icon sets more easily.
- Simplicity. We wouldn't need to overload our existing type definitions!

Cons:
- Cognitive overhead. This API isn't quite as ergonomic. It's also not how `<svg>` works on the web.
- Verbose. This API requires more typing.

**Comment by lloydjatkinson** [comment]
For reference this is the way I'm implementing SVG icons with Astro. It's pretty meh and long but works and also, of importance to me, has type safety on the icon names preventing typos. `IconType` is just a union of strings.

```ts
---
import type { IconType } from '../../domain/types';
import email from './email.svg?raw';
import twitter from './twitter.svg?raw';
import facebook from './facebook.svg?raw';
import linkedin from './linkedin.svg?raw';
import phosphorBooks from './phosphor-books.svg?raw';
import heroIconsCalculator from './heroicons-calculator.svg?raw';
import heroIconsCheck from './heroicons-check.svg?raw';
import heroIconsCode from './heroicons-code.svg?raw';
import heroIconsFeed from './heroicons-feed.svg?raw';
import heroIconsMoon from './heroicons-moon.svg?raw';
import heroIconsPlus from './heroicons-plus.svg?raw';
import heroIconsMinus from './heroicons-minus.svg?raw';
import heroIconsExclamationCircle from './heroicons-exclamation-circle.svg?raw';
import heroIconsChevronUp from './heroicons-chevron-up.svg?raw';
import heroIconsChevronDown from './heroicons-chevron-down.svg?raw';
import heroIconsEye from './heroicons-eye.svg?raw';
import heroIconsMagnifyingGlassPlus from './heroicons-magnifying-glass-plus.svg?raw';
import heroIconsMagnifyingGlassPlusMini from './heroicons-magnifying-glass-plus-mini.svg?raw';
import discord from './discord.svg?raw';
import envelope from './envelope.svg?raw';

export interface Props {
    name: IconType;
};
const { name } = Astro.props as Props;

const map: Record<IconType, string> = {
    'email': email,
    'twitter': twitter,
    'facebook': facebook,
    'linkedin': linkedin,
    'phosphor-books': phosphorBooks,
    'heroicons-calculator': heroIconsCalculator,
    'heroicons-check': heroIconsCheck,
    'heroicons-code': heroIconsCode,
    'heroicons-feed': heroIconsFeed,
    'heroicons-moon': heroIconsMoon,
    'heroicons-plus': heroIconsPlus,
    'heroicons-minus': heroIconsMinus,
    'heroicons-exclamation-circle': heroIconsExclamationCircle,
    'heroicons-chevron-up': heroIconsChevronUp,
    'heroicons-chevron-down': heroIconsChevronDown,
    'heroicons-eye': heroIconsEye,
    'heroicons-magnifying-glass-plus': heroIconsMagnifyingGlassPlus,
    'heroicons-magnifying-glass-plus-mini': heroIconsMagnifyingGlassPlusMini,
    'discord': discord,
    'envelope': envelope,
};

const svg = map[name];

// TODO: Size, colour, hover colour, etc.
---

<Fragment set:html={ svg } />
```

**Comment by jasikpark** [comment]
useful prior art in React-land is https://react-svgr.com/ - which I appreciate - allows creating a component per asset that can receive standard svg props and styling, inlining the SVG in the document

**Reply by jasikpark** [reply]
astro-svgr 🧐

**Reply by pReya** [reply]
And it's already available as a vite plugin: https://github.com/pd4d10/vite-plugin-svgr

**Comment by sannajammeh** [comment]
Working implementation on 3.0, works in 2.0 if you replace runtimePath with `astro/server/index.js`. 

> Important to note that this is my personal preferred implementation of this and I'm sharing this to provide an additional suggestion for a public API. I'm still very much in favor of the API's proposed by @natemoo-re .

Initial legwork was performed by a previous user and modified by me, however I can't find them or the original code anywhere 😳. Whoever wrote this, well done! 

### ⚠️⚠️ This is using internal imports from Astro - Not subject to semver - use at own risk ⚠️⚠️

## Vite Plugin

```ts
// svga.mjs
import { readFile } from "fs/promises";

const EXPORTS_DEFAULT = "export default ";

/**
  * @returns {import('vite').Plugin}
*/
export default function svga() {

  return {
    enforce: "pre",
    name: "astro:svga",

    async transform(code, id) {
      if (!id.endsWith(".svg?comp")) {
        return null;
      }
      const runtimePath = "astro/runtime/server/index.js";

      let idx = id;
      let codex = code;

      // remove the annoying querystring
      idx = id.replace("?comp", "");

      // this is because the `@astrojs/image` transforms the file already
      // and replaces with the url to load instead of the real source
      if (code.startsWith(EXPORTS_DEFAULT)) {
        codex = code.replace(EXPORTS_DEFAULT, "");
        codex = await readFile(idx, "utf-8");
      }

      const template = /*ts*/`import {
				render,
				createAstro,
				createComponent,
				spreadAttributes,
			  } from "${runtimePath.toString()}";
			  
			  const Astro = createAstro();
			  
			  const SVG = createComponent(async (result, props, slots) => {
				const _Astro = result.createAstro(Astro, props, slots);
				Astro.self = SVG;
				return render\`<svg \${spreadAttributes(_Astro.props, "props")}%%svg%%\`;
			  }, "%%id%%");
			  
			  ${EXPORTS_DEFAULT}SVG;
			  
			  if (import.meta.hot) import.meta.hot.decline()`;

      const resolved = template
        .replace("%%id%%", idx)
        // the `<svg` is already in the template since we want to spread Astro props.
        .replace("%%svg%%", codex.replace("<svg", ""));

      return {
        code: resolved,
        meta: {
          astro: {
            hydratedComponents: [],
            clientOnlyComponents: [],
            scripts: [],
            propagation: "none",
            containsHead: false,
            pageOptions: {},
          },
          vite: {
            lang: "ts",
          },
        },
      };
    },
  };
}
```

Also ship proper TS types for Astro imports:
```ts
// svga-env.d.ts
declare module "*.svg?comp" {
	import type { HTMLAttributes } from "astro/types";

	const component: (
		props: HTMLAttributes<"svg">,
	) => astroHTML.JSX.DefinedIntrinsicElements["svg"];

	export default component;
}
```

## Usage

```ts
// astro.config.mjs
  vite: {
    plugins: [svga()],
  },
  ...rest
```
In components use `?comp` to signal it should be imported as a component
```astro
---
import HorizontalPillar from "@/assets/fixtures/horizontal-pillar.svg?comp";
import VerticalPillar from "@/assets/fixtures/vertical-pillar.svg?comp";
---

 <HorizontalPillar class="absolute bottom-0 -left-4 max-w-full h-100px" />
  <VerticalPillar
    class="absolute bottom-0 right-0 -mb-3 pointer-events-none max-w-100px"
  />
```

**Reply by jasikpark** [reply]
Oooh Nice! `astro-svga` seems useful - you could also import a bunch of svgs that way and re-export them to create a sort of asset lib

**Reply by natemoo-re** [reply]
Nice! Absolutely love this. Super similar to [my working branch](https://github.com/withastro/astro/compare/main...spike/svg)!

**Comment by natemoo-re** [comment]
OK so I explicitly did NOT want to wade into the whole icon ecosystem thing initially, but I'm sure that's what most people are interested in.

Very curious if an import-like API that piggybacks on `astro:assets` is interesting to anyone? Very similar to [`unplugin-icons`](https://github.com/unplugin/unplugin-icons) but potentially built-in.

```astro
---
import Camera from 'astro:assets/icons/mdi/camera'
---

<Camera />
```

Here's a quick demo:

https://github.com/withastro/roadmap/assets/7118177/b6c221e8-ede4-4e86-94cc-4105d07a35c9

**Reply by itsmatteomanf** [reply]
That looks good, mentioned in Discord, but gonna keep it here, too.

Please support `class` (Tailwind...) and custom SVGs for paid or custom icons. Then it's gonna be the ultimate tool.

**Reply by natemoo-re** [reply]
Obviously types are a whole set of problems to solve, but we already do some typegen for `astro:content` so we definitely have a place to generate types.

**Reply by sannajammeh** [reply]
Now this looks amazing! Perhaps custom SVG's can be imported using `<path>:asset` or `<path>?asset` or maybe even a `(?/:)astro` to signal developers that it will be transformed to an astro component?

```ts
import Logo from "../assets/images/logo.svg?astro"; 
```

**Comment by natemoo-re** [comment]
Great news! This proposal has been accepted and is moving on to [Stage 2](https://github.com/withastro/roadmap#stage-2-accepted-proposal) of our process! 🎉

See [Built-in SVG components](https://github.com/withastro/roadmap/issues/699).

## What does that mean?

The Astro [maintainers](https://github.com/withastro/.github/blob/main/GOVERNANCE.md#level-2-l2---maintainer) and [technical steering committee](https://github.com/withastro/.github/blob/main/GOVERNANCE.md#technical-steering-committee-tsc) have agreed to **explore** official Astro support for this feature.

@natemoo-re has agreed to champion this proposal.

## What happens now?

- Our team will begin exploring possible approaches for supporting `.svg` components in Astro
- We will share progress updates [on the accepted proposal](https://github.com/withastro/roadmap/issues/000) as we progress
- **The goal of Stage 2 is to determine the feasibility of a proposed feature.** If it is feasible, the accepted proposal will graduate to an [Stage 3: RFC + Development](https://github.com/withastro/roadmap#stage-3-rfc--development) and implementation will begin

## Where can we follow along?

This discussion will be closed. Please follow along [on the accepted proposal](https://github.com/withastro/roadmap/issues/667). Don't hesitate to share additional feedback, use cases, or concerns [on the accepted proposal](https://github.com/withastro/roadmap/issues/667).

**Comment by NotWoods** [comment]
While the official feature is getting worked on, here's a quick workaround for anyone stumbling into this thread like I did. Tested in MDX post files.

```mdx
import svg from './image.svg?raw';

<Fragment set:html={svg} />
```

**Comment by cxa** [comment]


**Reply by itsmatteomanf** [reply]
It's been closed because it moved up a stage, per the comment by @natemoo-re above, https://github.com/withastro/roadmap/discussions/667#discussioncomment-6995376.

New link: https://github.com/withastro/roadmap/issues/699
